### PR TITLE
Include wcslib errors when parsing fits headers

### PR DESCRIPTION
**Pull request by aragilar**
It appears that errors produced when wcslib is scanning fits headers do not get passed up to the user. This adds them to the raised exception, as that appears to be the simplest way to pass them from C to python.

This is the first of the PRs that come from the discussion on slack around trying to read old INT files. This doesn't touch wcslib (only the astropy wrapper).

### CODE REVIEW

**Comment by aragilar** [comment]
Sure, happy to add docs/changelogs if they're desired. Not sure if its on the roadmap at some point, but rewriting the C wcslib wrapper in cython would make it easier to use tools like logging (and make it easier to test and contribute to).

**Comment by mcara** [comment]
@aragilar Could you, please, provide a link to a file that could be used for testing? Thanks!

**Comment by aragilar** [comment]
[mgc001_1_B.fit.gz](https://github.com/astropy/astropy/files/6985884/mgc001_1_B.fit.gz) is one of the bad files I have (this is from the MGC survey—Data Central has the originals, and will serve out fixed versions soon). You won't see the actual error (the use of ZPX rather than ZPN) with this PR, that needs some changes to wcslib (the wcslib changes need some cleaning up, I did some hacky stuff to work out why the file was invalid).

File: astropy/wcs/src/pyutil.c [hunk 1]
```diff
@@ -340,11 +340,21 @@ wcserr_fix_to_python_exc(const struct wcserr *err) {
 }
 
 void
-wcshdr_err_to_python_exc(int status) {
+wcshdr_err_to_python_exc(int status, const struct wcsprm *wcs) {
+  /* Add error to wcslib error buffer */
+  wcsperr(wcs, "");
```
Comment (mcara) [inline astropy/wcs/src/pyutil.c:345]:
```suggestion
  wcsperr(wcs, NULL);
```
From the [docs](https://www.atnf.csiro.au/people/mcalabre/WCS/wcslib/wcs_8h.html#a8fe5dcd9927240dc0348b850ee662367):
> prefix | If non-NULL, each output line will be prefixed with this string.

File: astropy/wcs/src/pyutil.c [hunk 2]
```diff
@@ -340,11 +340,21 @@ wcserr_fix_to_python_exc(const struct wcserr *err) {
...
   if (status > 0 && status != WCSHDRERR_PARSER) {
-    PyErr_SetString(PyExc_MemoryError, "Memory allocation error");
+    PyErr_Format(
+      PyExc_MemoryError,
+      "Memory allocation error (internal details %s)",
```
Comment (mcara) [inline astropy/wcs/src/pyutil.c:349]:
```suggestion
      "Memory allocation error:\n%s",
```

File: astropy/wcs/src/pyutil.c [hunk 3]
```diff
@@ -340,11 +340,21 @@ wcserr_fix_to_python_exc(const struct wcserr *err) {
...
+      wcsprintf_buf()
+    );
   } else {
-    PyErr_SetString(PyExc_ValueError, "Internal error in wcslib header parser");
+    PyErr_Format(
+      PyExc_ValueError,
+      "Internal error in wcslib header parser (internal details %s)",
```
Comment (mcara) [inline astropy/wcs/src/pyutil.c:355]:
```suggestion
      "Internal error in wcslib header parser:\n %s",
```

**Review by mcara — CHANGES_REQUESTED** [review]
Seems reasonable. I added a couple of minor suggestions that surrounding the original error message with parenthesis does not look clean to me. For example, currently the error message looks like this:
```
ValueError: Internal error in wcslib header parser (internal details EQUINOX = '2000.   '           / Equinox of coordinates
  Rejected (1), a floating-point value was expected.
RADECSYS= 'FK5     '           / mean place new (after the 1976 IAU) system
  Rejected (2), the RADECSYS keyword is deprecated, use RADESYSa.
)
```
I think something like this:
```
ValueError: Internal error in wcslib header parser:
EQUINOX = '2000.   '           / Equinox of coordinates
  Rejected (1), a floating-point value was expected.
RADECSYS= 'FK5     '           / mean place new (after the 1976 IAU) system
  Rejected (2), the RADECSYS keyword is deprecated, use RADESYSa.
```
is somewhat better looking.

**Comment by aragilar** [comment]
@mcara I've added your suggestions around the text, I agree they look better.

**Review by mcara — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**